libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/building/BuildingProcessor.kt

```diff
@@ -0,0 +1,58 @@
+package com.mapbox.navigation.ui.maps.building
+
+import com.mapbox.bindgen.ExpectedFactory
+import com.mapbox.maps.RenderedQueryOptions
+import com.mapbox.navigation.ui.maps.building.model.BuildingError
+import com.mapbox.navigation.ui.maps.building.model.BuildingValue
+import kotlin.coroutines.resume
+import kotlin.coroutines.suspendCoroutine
+
+internal object BuildingProcessor {
+
+    /**
+     * Source layer that includes building height.
+     */
+    private const val BUILDING_LAYER_ID = "building"
+
+    suspend fun queryBuilding(
+        action: BuildingAction.QueryBuilding
+    ): BuildingResult.QueriedBuildings = suspendCoroutine { continuation ->
+        val screenCoordinate = action.mapboxMap.pixelForCoordinate(action.point)
+        val queryOptions = RenderedQueryOptions(listOf(BUILDING_LAYER_ID), null)
+        action.mapboxMap.queryRenderedFeatures(screenCoordinate, queryOptions) { expected ->
+            expected.fold(
+                { error ->
+                    continuation.resume(
+                        BuildingResult.QueriedBuildings(
+                            ExpectedFactory.createError(BuildingError(error))
+                        )
+                    )
+                },
+                { value ->
+                    continuation.resume(
+                        BuildingResult.QueriedBuildings(
+                            ExpectedFactory.createValue(BuildingValue(value))
+                        )
+                    )
+                }
+            )
+        }
+    }
+
+    fun queryBuildingOnWaypoint(
+        action: BuildingAction.QueryBuildingOnWaypoint
+    ): BuildingResult.GetDestination {
+        val routeOptions = action.progress.route.routeOptions()
+        val coordinateIndex = action.progress.currentLegProgress?.legIndex!! + 1
+        return routeOptions?.coordinatesList()?.getOrNull(coordinateIndex)?.let { point ->
+            BuildingResult.GetDestination(point)
+        } ?: BuildingResult.GetDestination(null)
```

Comment:
Is null a valid result or should we report an error here?

Reply:
I am returning `null` to the callee in this case the `MapboxBuildingsApi.queryBuildingOnWaypoint()` who is then returning an error. If not do this then, I would have to add another state to the sealed class `BuildingDestination`.
Instead of 
```
sealed class BuildingResult {
    sealed class BuildingDestination(point: Point?): BuildingResult()
}
```

this would have to be changed to 
```
sealed class BuildingResult {
    sealed class BuildingDestination: BuildingResult() {
        data class Error(msg: String)
        data class Success(point: Point)
    }
}
```

Reply:
I see :+1: Might be clearer to have a separate state rather than if-checking in the API, not a blocker though.